library/data/tags/src/main/java/org/quiltmc/qsl/tag/impl/TagRegistryImpl.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2016, 2017, 2018, 2019 FabricMC, 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.tag.impl;
+
+import java.util.function.Supplier;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.resource.ResourceManager;
+import net.minecraft.resource.ServerResourceManager;
+import net.minecraft.server.Main;
+import net.minecraft.tag.RequiredTagList;
+import net.minecraft.tag.RequiredTagListRegistry;
+import net.minecraft.tag.ServerTagManagerHolder;
+import net.minecraft.tag.Tag;
+import net.minecraft.tag.TagGroup;
+import net.minecraft.tag.TagGroupLoader;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.dynamic.RegistryOps;
+import net.minecraft.util.registry.DynamicRegistryManager;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+
+import org.quiltmc.qsl.tag.api.TagRegistry;
+import org.quiltmc.qsl.tag.api.TagType;
+import org.quiltmc.qsl.tag.impl.client.ClientTagRegistryManager;
+import org.quiltmc.qsl.tag.mixin.DynamicRegistryManagerAccessor;
+import org.quiltmc.qsl.tag.mixin.RequiredTagListRegistryAccessor;
+
+@SuppressWarnings("ClassCanBeRecord")
+@ApiStatus.Internal
+public final class TagRegistryImpl<T> implements TagRegistry<T> {
+	private static final Logger LOGGER = LogManager.getLogger();
+	private static final ThreadLocal<Boolean> MISSING_TAGS_CLIENT_FETCH = new ThreadLocal<>();
```

Comment:
Is a thread local an error here? My interpretation of the uses is that the integrated server would never see the client thread signal it that it is fetching tags because the signal would never signal the server's thread.

If it is an error, would `AtomicBoolean` suffice.

Reply:
This boolean is used exclusively on the client to check whether there is missing tags that should prevent the connection to the server, the entire goal of this ThreadLocal is to change the behavior of `getMissingTags` to exclude any tags not required to connect in `onSynchronizeTags(SynchronizeTagsS2CPacket)`.

`AtomicBoolean` could introduce unwanted behavior on the integrated server.